rust/kernel/thread.rs

```diff
@@ -0,0 +1,508 @@
+    }
+
+    /// Wakes up the thread.
+    ///
+    /// Note that a newly created thread (e.g. via [`Thread::try_new`]) will not
+    /// run until a [`Thread::wake_up`] is called.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep, don't call it in atomic contexts.
+    pub fn wake_up(&self) {
+        // SAFETY: `task` is a valid pointer to a kernel thread structure, the refcount
+        // of which is increased in `try_new*`, so it won't point to a freed
+        // `task_struct`. And it's not stopped because `stop` will consume the
+        // [`Thread`].
+        unsafe {
+            bindings::wake_up_process(self.task);
+        }
+    }
+
+    /// Stops the thread.
+    ///
+    /// - If the thread hasn't been waken up after creation, the thread closure
+    ///   won't be called, and will return `-EINTR`. Note that a thread may not
+    ///   be waken up even after [`Thread::wake_up`] is called.
+    ///
+    /// - Otherwise, waits for the closure to return or the thread `do_exit`
+    ///   itself.
+    ///
+    /// Consumes the [`Thread`] so that it's not accessible. In case of error,
+    /// returns the exit code of the thread.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep, don't call it in atomic contexts.
+    pub fn stop(self) -> KernelResult<()> {
+        // SAFETY: `task` is a valid pointer to a kernel thread structure, the
+        // refcount of which is increased in `try_new*`, so it won't point to
+        // a freed `task_struct`. And it's not stopped because `stop` will
+        // consume the [`Thread`].
+        let ret = unsafe { bindings::kthread_stop(self.task) };
+
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(Error::from_kernel_errno(ret))
+        }
+    }
+}
+
+impl Drop for Thread {
+    fn drop(&mut self) {
+        // Decreases the refcount of the thread, the thread may still be running after
+        // we `drop` the `Thread`.
+        //
+        // SAFETY: At least one refcount is held by `Thread::try_new*` and
+        // the refcount of `task_struct` is implemented by atomics.
+        unsafe {
+            rust_helper_put_task_struct(self.task);
+        }
+    }
+}
+
+/// Tries to give up the CPU and lets another thread run.
+///
+/// This maps to kernel's `schedule` function, which is similar to
+/// [`std::thread::yield_now`].
+///
+/// # Context
+///
+/// This function might sleep, don't call in atomic contexts.
+///
+/// [`std::thread::yield_now`]: https://doc.rust-lang.org/std/thread/fn.yield_now.html
+#[doc(alias = "yield_now")]
+pub fn schedule() {
+    // SAFETY: ...
+    // If we can schedule back from other thread, then this can be treated as a
+    // no-op. A special case occurs when a thread sets its state to `TASK_DEAD`,
+    // and then `schedule` will not come. Currently we don't have a way to do
+    // this safely in Rust, and in the future, we probably still won't allow it.
```

Comment:
Oh -- when I wrote `// SAFETY: ...` in one of the suggestions before, I meant to put the paragraph right there (i.e. without the newlines), rather than the actual `...`! In other words, something like this:

```suggestion
    // SAFETY: If we can schedule back from other thread, then this can be treated as a
    // no-op. A special case occurs when a thread sets its state to `TASK_DEAD`,
    // and then `schedule` will not come. Currently we don't have a way to do
    // this safely in Rust, and in the future, we probably still won't allow it.
```

Reply:
Ok, got it. I will fix this ;-)